File: src/_004_interview_preparation/_03_leetcode/_grind75/_02_valid_parentheses/ValidParentheses.java
package _004_interview_preparation._03_leetcode._grind75._02_valid_parentheses;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

public class ValidParentheses {

    public static void main(String[] args) {
        String s1 = "()";
        String s2 = "()[]{}";
        String s3 = "(]";
        String s4 = "[[(())]]";

        System.out.println("Test 1: () = " + isValid(s1));
        System.out.println("Test 2: ()[]{} = " + isValid(s2));
        System.out.println("Test 3: (] = " + isValid(s3));
        System.out.println("Test 4: [[(())]] = " + isValid(s4));
    }

    public static boolean isValid(String s) {

        /*Map<Character, Character> bracketMap = new HashMap<>();
        bracketMap.put(')', '(');
        bracketMap.put('}', '{');
        bracketMap.put(']', '[');*/

        Map<Character, Character> bracketMap = Map.of(')', '(', '}', '{', ']', '[');

        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (bracketMap.containsValue(c)) { // opening bracket
                stack.push(c);
            } else if (bracketMap.containsKey(c)) { // closing bracket
                if (stack.empty() || stack.pop() != bracketMap.get(c)) {
                    return false;
                }
            }
        }

        return stack.empty();
    }
}

File: src/_004_interview_preparation/_03_leetcode/_grind75/_02_valid_parentheses/ValidParentheses.kt
package _004_interview_preparation._03_leetcode._grind75._02_valid_parentheses

import java.util.Stack

fun main() {
    val solution = Solution()

    val s1 = "()"
    val s2 = "()[]{}"
    val s3 = "(]"
    val s4 = "[[(())]]"

    println("Test1 () = ${solution.isValid(s1)}")
    println("Test2 ()[]{} = ${solution.isValid(s2)}")
    println("Test3 (] = ${solution.isValid(s3)}")
    println("Test4 [[(())]] = ${solution.isValid(s4)}")
}

class Solution {
    fun isValid(s: String): Boolean {
        val bracketMap = mapOf(')' to '(', ']' to '[', '}' to '{')

        val stack = Stack<Char>()

        for (c in s) {
            if (bracketMap.containsValue(c)) {
                stack.push(c);
            } else if (bracketMap.containsKey(c)) {
                if (stack.empty() || stack.pop() != bracketMap[c]) {
                    return false
                }
            }
        }
        return stack.empty()
    }
}